### PR TITLE
Update API

### PR DESCRIPTION
**Pull request by demurgos**
This PR is part of the effort to improve Node's support for code coverage (see discussion on `c8`).

This PR exposes two new functions to support foreground childs:
- `proxy(parent, child)` to start forwarding between two existing processes.
- `spawn(file, args, options)` as a replacement for the original functions allowing you to pass `spawn` options.

The original API is still there. The only breaking change is dropping support for unmaintained Node versions (require Node 6+).

The new `spawn` function would allow to [pass the env NODE_V8_COVERAGE variable directly in c8](https://github.com/bcoe/c8/blob/6cf360197e6887180be4a45c471fad35f669de68/bin/c8.js#L25).
There are a few other changes:
- It always defaults to use the builtin `cp.spawn` function (even on Windows), you can pass your own function if you want to.
- It does not take over control flow: it is your responsibility to exit the parent process once the child is closed.

Other comments:
- I converted the lib to Typescript (for documentation). I also converted the tests to make sure the types work fine.
- I completed the test suite to reach 100% coverage.
- From a semver point of view, this is a major version.
- I update the dependencies.

Discussions:
- Maybe provide a way to automatically kill the parent process like in the original function?
- Migrating from the original function to `fg.spawn` should be relatively easy. For the moment I marked it as `deprecated`, I don't think it should be removed immediately (especially since it's known to work fine with `nyc`). Not sure what you think about it.

- Closes tapjs/foreground-child#13
- Closes tapjs/foreground-child#14
- Closes tapjs/foreground-child#15
- Closes tapjs/foreground-child#16
- Closes tapjs/foreground-child#17
- Closes tapjs/foreground-child#18
- Closes tapjs/foreground-child#19

### CODE REVIEW

**Comment by demurgos** [comment]
Closing for now, I'll send smaller PRs.

File: README.md [hunk 1]
```diff
@@ -2,34 +2,34 @@
...
-```
 
-A callback can optionally be provided, if you want to perform an action
-before your foreground-child exits:
+// Close is a promise resolved when the child is closed.
+// You can use it to inspected its exit code and exit the parent process.
+close.then((close) => {
+  // You can use this handler to perform an action before exiting the
+  // foreground child.
+  console.log(close.code);
```
Comment (profnandaa) [inline README.md:28]: How's `close` a function and object at same time?

File: appveyor.yml [hunk 2]
```diff
@@ -1,8 +1,8 @@
 environment:
   matrix:
-    - nodejs_version: '5'
-    - nodejs_version: '4'
-    - nodejs_version: '0.12'
```
Comment (profnandaa) [inline appveyor.yml:5]: What's the rationale for dropping support for <6?

File: custom-typings/signal-exit/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+declare module "signal-exit" {
```
Comment (profnandaa) [inline custom-typings/signal-exit/index.d.ts:1]: From  our discussion in the channel, I think it's better off dropping off TS from this PR, for now.

File: test/immortal-child.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
+import t from "tap";
```
Comment (profnandaa) [inline test/immortal-child.ts:1]: Separate out increasing test-coverage as a different PR, but only have tests for the code you've written for the PR.

File: custom-typings/signal-exit/index.d.ts (hunk 3 above)
Comment (profnandaa) [reply custom-typings/signal-exit/index.d.ts:1]: Rewrite this in JS (even not ES6 for now), for backward compatibility.

**Review by profnandaa — COMMENTED** [review]
Please have a look in consideration for your rewrite.

File: README.md (hunk 1 above)
Comment (demurgos) [reply README.md:28]: Well, it's like most CJS exports: a function with additional properties (`fg()` and `fg.spawn()`, `signalExit()` and `signalExit.signals()`, etc.).

It boils down to:
```javascript
const close = () => { ... };
close.code = code;
close.signal = signal;
return close;
```

File: appveyor.yml (hunk 2 above)
Comment (demurgos) [reply appveyor.yml:5]: - Align with Node's support policy by targeting oldest LTS.
- Use promise-based API out-of-the box (without having to use `bluebird`)
- Remove workarounds for old Node versions
- Refactor the internals to use ES syntax (scoped variables, spread args, arrow functions, etc.)

File: test/immortal-child.ts (hunk 4 above)
Comment (demurgos) [reply test/immortal-child.ts:1]: OK

File: README.md (hunk 1 above)
Comment (profnandaa) [reply README.md:28]: Oh, got it. Thanks!

File: appveyor.yml (hunk 2 above)
Comment (profnandaa) [reply appveyor.yml:5]: I see. Looks that just this alone should be a different PR/discussion with the maintainers if they are aligned. That way, it becomes more progressive than big-bang.

**Pull request closed without merging**